src/xdp/apikernel.c

```diff
@@ -0,0 +1,517 @@
+    HANDLE NmrProviderHandle;
+} XDPAPI_PROVIDER;
+
+typedef struct _XDPAPI_CLIENT {
+    HANDLE NmrBindingHandle;
+    GUID ClientModuleId;
+    const XDP_API_CLIENT_DISPATCH *ClientDispatch;
+    const VOID *ClientBindingContext;
+} XDPAPI_CLIENT;
+
+typedef struct _XDP_API_ROUTINE {
+    _Null_terminated_ const CHAR *RoutineName;
+    VOID *Routine;
+} XDP_API_ROUTINE;
+
+static XDPAPI_PROVIDER XdpApiProvider;
+static CONST NPI_MODULEID NPI_XDP_MODULEID = {
+    sizeof(NPI_MODULEID),
+    MIT_GUID,
+    { 0x521421C2, 0x539B, 0x46DF, { 0xA1, 0x43, 0x8D, 0xAC, 0x88, 0xD0, 0xC1, 0x55 } }
+};
+
+//
+// API routines.
+//
+static XDP_GET_ROUTINE_FN XdpApiKernelXdpGetRoutine;
+static XDP_CREATE_PROGRAM_FN XdpApiKernelXdpCreateProgram;
+static XDP_DELETE_PROGRAM_FN XdpApiKernelXdpDeleteProgram;
+static XDP_INTERFACE_OPEN_FN XdpApiKernelXdpInterfaceOpen;
+static XDP_INTERFACE_CLOSE_FN XdpApiKernelXdpInterfaceClose;
+static XSK_CREATE_FN XdpApiKernelXskCreate;
+static XSK_DELETE_FN XdpApiKernelXskDelete;
+static XSK_BIND_FN XdpApiKernelXskBind;
+static XSK_ACTIVATE_FN XdpApiKernelXskActivate;
+static XSK_NOTIFY_SOCKET_FN XdpApiKernelXskNotifySocket;
+static XSK_NOTIFY_ASYNC2_FN XdpApiKernelXskNotifyAsync2;
+static XSK_SET_SOCKOPT_FN XdpApiKernelXskSetSockopt;
+static XSK_GET_SOCKOPT_FN XdpApiKernelXskGetSockopt;
+static XSK_IOCTL_FN XdpApiKernelXskIoctl;
+
+//
+// Experimental APIs, subject to removal in a minor release.
+//
+static XDP_RSS_GET_CAPABILITIES_FN XdpApiKernelXdpRssGetCapabilities;
+static XDP_RSS_SET_FN XdpApiKernelXdpRssSet;
+static XDP_RSS_GET_FN XdpApiKernelXdpRssGet;
+static XDP_QEO_SET_FN XdpApiKernelXdpQeoSet;
+
+#define DECLARE_XDP_API_ROUTINE(_routine) #_routine, (VOID *)XdpApiKernel##_routine
+#define DECLARE_EXPERIMENTAL_XDP_API_ROUTINE(_routine, _name) _name, (VOID *)XdpApiKernel##_routine
+
+static const XDP_API_ROUTINE XdpApiRoutines[] = {
+    { DECLARE_XDP_API_ROUTINE(XdpGetRoutine) },
+    { DECLARE_XDP_API_ROUTINE(XdpCreateProgram) },
+    { DECLARE_XDP_API_ROUTINE(XdpDeleteProgram) },
+    { DECLARE_XDP_API_ROUTINE(XdpInterfaceOpen) },
+    { DECLARE_XDP_API_ROUTINE(XdpInterfaceClose) },
+    { DECLARE_XDP_API_ROUTINE(XskCreate) },
+    { DECLARE_XDP_API_ROUTINE(XskDelete) },
+    { DECLARE_XDP_API_ROUTINE(XskBind) },
+    { DECLARE_XDP_API_ROUTINE(XskActivate) },
+    { DECLARE_XDP_API_ROUTINE(XskNotifySocket) },
+    { DECLARE_XDP_API_ROUTINE(XskNotifyAsync2) },
+    { DECLARE_XDP_API_ROUTINE(XskSetSockopt) },
+    { DECLARE_XDP_API_ROUTINE(XskGetSockopt) },
+    { DECLARE_XDP_API_ROUTINE(XskIoctl) },
+    { DECLARE_EXPERIMENTAL_XDP_API_ROUTINE(XdpRssGetCapabilities, XDP_RSS_GET_CAPABILITIES_FN_NAME) },
+    { DECLARE_EXPERIMENTAL_XDP_API_ROUTINE(XdpRssSet, XDP_RSS_SET_FN_NAME) },
+    { DECLARE_EXPERIMENTAL_XDP_API_ROUTINE(XdpRssGet, XDP_RSS_GET_FN_NAME) },
+    { DECLARE_EXPERIMENTAL_XDP_API_ROUTINE(XdpQeoSet, XDP_QEO_SET_FN_NAME) },
+};
+
+static CONST XDP_API_PROVIDER_DISPATCH XdpApiProviderDispatchV1 = {
```

Comment:
Can we use `const` everywhere instead of `CONST` (like you do above)?

P.S. I really don't like the ALL CAPS types of Windows, and would like to replace those with the normal C types eventually...

Reply:
Yeah, I've been working to use `const` rather than `CONST` within newer code in XDP, but the old stuff lingers. Might be worth a quick search/replace in a standalone PR to clean up the stragglers.

Reply:
Addressed with #453